add docs for remote write (#83)
pyroscope-io/docs: docs/remote-write.mdx

Its current implementation is simple: for all ingested data, the same data is sent to the remote targets' `/ingest` endpoint, concurrently.

:::note
We currently do not provide any guarantees this data is going to be delivered (think for example, network failures). Although in future versions we may implement retries.

:::

## Tuning

Although we try to provide [smart defaults](https://github.com/pyroscope-io/pyroscope/blob/12cf3a975215b9cb8b5c14d75b89437140c79284/pkg/remotewrite/concurrency.go#L9), not every environment is the same. Therefore we also allow a few configurable points:

```yaml
remote-write:
  enabled: true
  targets:
    bar:
      address: http://bar:4040
      auth-token: MY_AUTH_TOKEN
      tags:
        foo: bar

      # tweakable
      timeout: 1m
      queue-size: 100
      queue-workers: 4
```

The fields should be self-explanatory, but for completeness they are:

| Field | Description |
|--------|-------|
| `timeout`   | Maximum amount of time each upload can take. It uses the standard [Go time duration format](https://pkg.go.dev/time#ParseDuration). |
| `queue-size`       | Maximum amount of items that can be stored in the queue at a time. |

> | `queue-workers`       | How many queue workers are running. |

Author: I was thinking we would still tweak them, but if you feel strongly about it let's add them!

> | `queue-workers`       | How many queue workers are running. |

Reviewer: I agree even if we tweak them we should add them for now and update if needed

### Metrics
Prometheus metrics are also exposed:

| Metric | Type | Description  |
|--------|-------|---|
| `pyroscope_remote_write_client_response_time`   | histogram     |   |
| `pyroscope_remote_write_client_sent_bytes`       | counter      | The total number of bytes of data (not metadata) sent to the remote target.  |
| `pyroscope_remote_write_queue_capacity`       | gauge      | How many items the queue can hold. |
| `pyroscope_remote_write_queue_dropped` | counter      | How many items were dropped (as in not accepted into the queue). |
| `pyroscope_remote_write_queue_workers_total` | gauge | Total number of queue workers. |

Also notice that the key in the yaml is used as `targetName` in the metric. For example, given the following configuration:
```yaml
remote-write:
  enabled: true
  targets:
    my-cluster:
      address: http://bar:4040
```

> It will create metrics with the following labels (`targetName/targetAddress`):

Reviewer 2: Not related to the documentation: we already use **snake_case** for labels, maybe we should rename targetName and targetAddress?

Author: good point, added here https://github.com/pyroscope-io/pyroscope/pull/1157

```
pyroscope_remote_write_queue_pending{targetAddress="http://bar:4040",targetName="my-cluster"} 0
```

### Other Considerations
* Enabling remote write will increase bandwith usage, since every ingestion is replicated to the number of targets.

> * Also memory usage is increased, since data is replicated in memory to the number of targets.

Author: It depends somewhat (more targets means more changes that profile is still in memory), but I see what you mean, definitely not relevant enough to mention it

## What are the use cases?
There are mainly 2 use cases:

### Case 1: Running Pyroscope at the "edge"
Consider the following situation in an organization:
* multiple Kubernetes clusters
* [running pull mode](https://pyroscope.io/docs/golang-pull-mode/), ie. Pyroscope scrapes apps
* a central Pyroscope instance that these other remote pyroscope instances write to

That enables not exposing these internal Pyroscope instances, since they only scrape data. Therefore the only exposed instance is the central one.

### Case 2: migrating to Pyroscope Cloud
Consider a user that is running Pyroscope Open Source, and wants to experiment with the [Pyroscope Cloud version](https://pyroscope.cloud) (or any other implementation of the `/ingest` endpoint, as for example a Pyroscope Canary instance).

It requires updating all apps to point to Pyroscope Cloud, regenerating all API Keys, redeploying all apps, etc. Which is cumbersome.

Which remote write one only needs to add the Pyroscope Cloud instance as a target. For example:
```yaml
remote-write:

>   enabled: true

Reviewer 3: I would add
```
  disable-local-writes: true
```

here, because most people migrating to the cloud would want to disable local writes I think

  targets:

>     mycluster:

Reviewer 3: ```suggestion
    pyroscope-cloud:
```

>       address: https://ingest.pyroscope.io

Reviewer 3: ```suggestion
      address: https://ingest.pyroscope.cloud
```

      auth-token: MY_AUTH_TOKEN
      tags:

>         cluster-name: my-cluster

Reviewer 3: ```suggestion
        cluster-name: my-cluster # optional tag to distinguish between different clusters you run
```

---
id: remote-write
title: Remote Write
sidebar_label: Remote Write
slug: /remote-write
description: Push data to another pyroscope instance
keywords: [ pyroscope, server, remote, write, storage ]
---

:::note
This functionality is available starting from [pyroscope version 0.19.0](https://github.com/pyroscope-io/pyroscope/releases/tag/v0.19.0).

Make sure to upgrade before you use this.
:::

## What is Remote Write?
Remote Write is a feature that allows writing data to a remote Pyroscope instance (ie. not the same one that is immediately ingesting the data).

## How does it work?
<!-- TODO: create a diagram? -->

Its current implementation is simple: for all ingested data, the same data is sent to the remote targets' `/ingest` endpoint, concurrently.

:::note
We currently do not provide any guarantees this data is going to be delivered (think for example, network failures). Although in future versions we may implement retries.
:::

## Basic Example
```yaml
remote-write:
  enabled: true
  targets:
    bar:
      address: http://bar:4040

>       auth-token: MY_AUTH_TOKEN

Reviewer: don't feel strongly about this, but I suppose foo and bar will have different auth tokens 
```suggestion
      auth-token: MY_BAR_AUTH_TOKEN
```

      tags:
        foo: bar
    baz:
      address: http://baz:4040

>       auth-token: MY_AUTH_TOKEN

Reviewer: don't feel strongly about this, but I suppose foo and bar will have different auth tokens 
```suggestion
      auth-token: MY_BAZ_AUTH_TOKEN
```

      tags:
        foo: baz
```

> For every single ingested profile, it will also write to 2 pyroscope instances: `http://bar:4040` and `http://baz:4040`.

Reviewer: ```suggestion
For every ingested profile, it will also write to 2 pyroscope instances: `http://bar:4040` and `http://baz:4040`.
```

[Tags](https://pyroscope.io/blog/profiling-go-apps-with-pyroscope/#narrowing-in-on-the-issue-using-tags) can (and should) be added, so that you can identify where the profiling data is coming from.

## What are the use cases?
There are mainly 2 use cases:

> ### Case 1: Running Pyroscope at the "edge"

Reviewer 3: Centralizing -> Aggregating

Consider the following situation in an organization:
* multiple Kubernetes clusters
* [running pull mode](https://pyroscope.io/docs/golang-pull-mode/), ie. Pyroscope scrapes apps
* a central Pyroscope instance that these other remote pyroscope instances write to

That enables not exposing these internal Pyroscope instances, since they only scrape data. Therefore the only exposed instance is the central one.

### Case 2: migrating to Pyroscope Cloud

> Consider a user that is running Pyroscope Open Source, and wants to experiment with the [Pyroscope Cloud version](https://pyroscope.cloud) (or any other implementation of the `/ingest` endpoint, as for example a Pyroscope Canary instance).

Reviewer: ```suggestion
For users who are running Pyroscope's Open Source version and would like to migrate to Pyroscope's Cloud version, _before_ `remote-write` it required:

1. Updating applications to point to Pyroscope Cloud
2. Replacing API Keys
3. Redeploying applications

:::note
If you'd like to join the closed beta for Pyroscope cloud, please reach out to support@pyroscope.io and we'll work with you to get you started!
:::
```

> It requires updating all apps to point to Pyroscope Cloud, regenerating all API Keys, redeploying all apps, etc. Which is cumbersome.

Reviewer: Removing in favor of comment above where this was merged in
```suggestion
```

> With remote write one only needs to add the Pyroscope Cloud instance as a target. For example:

Reviewer: ```suggestion
With remote write, migrating to the cloud is simple as adding the Pyroscope Cloud instance as a target (and as you transition your open source data will be unaffected)! For example:
```

```yaml
remote-write:
  enabled: true
  disable-local-writes: true
  targets:
    pyroscope-cloud:
      address: https://ingest.pyroscope.cloud
      auth-token: MY_AUTH_TOKEN
      tags:
        # optional tag to distinguish between different clusters you have
        cluster-name: my-cluster
```
## Disabling local writes
The `disable-local-writes` configuration allows disabling writes to the local storage.

> It's useful if you need just to relay data to another pyroscope instance.

Reviewer: ```suggestion
We recommend using this option if you only want to relay data to another pyroscope instance and don't need to write it locally.
```

> Keep in mind it may be useful to leave local writes enabled, so that if the remote target is having issues, it's still possible to access the ingested data.

Reviewer: ```suggestion
Keep in mind, in some situations it may be useful to leave local writes enabled, so that if a remote target is having issues, it's still possible to access the ingested data locally.
```

```yaml
remote-write:
  enabled: true
  disable-local-writes: true
```
## Tuning

> Although we try to provide [smart defaults](https://github.com/pyroscope-io/pyroscope/blob/12cf3a975215b9cb8b5c14d75b89437140c79284/pkg/remotewrite/concurrency.go#L9), not every environment is the same. Therefore we also allow a few configurable points:

Reviewer: ```suggestion
 We try recommend using the following [defaults](https://github.com/pyroscope-io/pyroscope/blob/12cf3a975215b9cb8b5c14d75b89437140c79284/pkg/remotewrite/concurrency.go#L9), however, not every environment is the same. If you have trouble configuring default values please reach out to support@pyroscope.io or join our [slack channel](https://pyroscope.io/slack) and we'd love to help. 
 
 That being said, Pyroscope has a few key configurable options:
```

Reviewer 3: We recommend

```yaml
remote-write:
  enabled: true
  targets:
    bar:
      address: http://bar:4040
      auth-token: MY_AUTH_TOKEN
      tags:
        foo: bar

      # tweakable
      timeout: 1m
      queue-size: 100
      queue-workers: 4
```

> The fields should be self-explanatory, but for completeness they are:

Reviewer: I think typically calling things "self-explanatory" only risks people feeling like they are not so and feeling bad if they don't understand
```suggestion
```
